### PR TITLE
resourcemanager: scheduler subtask in the pool's task

### LINKED ISSUE #40719: resourcemanager: scheduler subtask in the pool's task
## Enhancement

Now, we can control the concurrency of the pool. But we need to wait for the subtask to finish when we need to downclock the pool, then we can exit the worker in the pool. if we want to overclock the pool, we need to create a subtask for adding concurrency to the task. so we have the ability to control the subtask to make itself exiting or creating.

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: close #40719

Problem Summary:

### What is changed and how it works?

Now, we can control the concurrency of the pool. But we need to wait for the subtask to finish when we need to downclock the pool, then we can exit the worker in the pool. if we want to overclock the pool, we need to create a subtask for adding concurrency to the task. so we have the ability to control the subtask to make itself exiting or creating.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by YangKeao — APPROVED** [review]
LGTM

**Review by Defined2014 — APPROVED** [review]

File: resourcemanager/pooltask/task_manager_iterator.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
+// Copyright 2022 PingCAP, Inc.
```
Comment (zimulala) [inline resourcemanager/pooltask/task_manager_iterator.go:1]:
```suggestion
// Copyright 2023 PingCAP, Inc.
```

Comment (zimulala) [reply resourcemanager/pooltask/task_manager_iterator.go:1]: Please add UTs for this file.

Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager_iterator.go]: This test is in ```util/gpool/spmc/spmcpool_test.go```.

Comment (zimulala) [reply resourcemanager/pooltask/task_manager_iterator.go]: I feel that it is an independent library and should have its own UTs.

Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager_iterator.go]: In fact, it is convenient for us to write a unit test for this at the pool. But I find we cannot see the UT coverage in the PR. I have reported it to EE team.

File: resourcemanager/pooltask/task_manager_iterator.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		box := findTask[T, U, C, CT, TF](m, RunningTask)
+		if box != nil {
+			return box, false
+		}
+	}
+	return nil, false
+}
+
+func canBoost[T any, U any, C any, CT any, TF Context[CT]](m *meta[T, U, C, CT, TF], min time.Time) (result *list.Element, isBreak bool) {
+	if m.running.Load() < m.origin {
```
Comment (zimulala) [inline resourcemanager/pooltask/task_manager_iterator.go:100]: Could we rename the field `m.origin`? It is hard to understand, maybe like `originalConcurrency` or `initialConcurrency` or something more readable.

File: resourcemanager/pooltask/task_manager_iterator.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+			t.task[i].rw.RLock()
+			defer t.task[i].rw.RUnlock()
+			for id, stats := range t.task[i].stats {
+				if result == nil {
+					result = findTask[T, U, C, CT, TF](stats, RunningTask)
+					tid = id
+					compareTS = stats.createTS
+					continue
+				}
+				newResult, isBoost := fn(stats, compareTS)
```
Comment (zimulala) [inline resourcemanager/pooltask/task_manager_iterator.go:61]: This naming(`isBoost`) doesn't feel right, "canPause" should return `isPause`?

Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager_iterator.go]: ```iter``` is used by ```overclock``` and ```downclock```. so I rename it ```breakFind```.

File: resourcemanager/pooltask/task_manager_iterator.go (hunk 2 above)
Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager_iterator.go]: Done

**Comment by hawkingrei** [comment]
@zimulala PTAL

File: resourcemanager/pooltask/task_manager_iterator.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+			for id, stats := range t.task[i].stats {
+				if result == nil {
+					result = findTask[T, U, C, CT, TF](stats, RunningTask)
+					tid = id
+					compareTS = stats.createTS
+					continue
+				}
+				newResult, pauseFind := fn(stats, compareTS)
+				if pauseFind {
+					result = newResult
```
Comment (zimulala) [inline resourcemanager/pooltask/task_manager_iterator.go:63]: If here we call `canBoost`, and we return `nil, true` https://github.com/pingcap/tidb/pull/40670/files#diff-f1e67ff3fe9cec68e358a2e0ab7bc91ecca0dfcee2e9827282a0b6c060449a64R101. Then `newResult` is nil, is this expected and why？

**Review by zimulala — APPROVED** [review]

File: resourcemanager/pooltask/task_manager_iterator.go (hunk 4 above)
Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager_iterator.go:63]: Fixed, it is a misjudgment.

**Pull request merged**